crates/viewer/re_renderer/src/renderer/video.rs

```diff
@@ -0,0 +1,544 @@
+    #[allow(clippy::undocumented_unsafe_blocks)]
+    unsafe impl Send for VideoFrame {}
+
+    /// # Safety
+    /// Wasm is single-threaded, so Send/Sync is safe
+    #[allow(unsafe_code)]
+    #[allow(clippy::undocumented_unsafe_blocks)]
+    unsafe impl Sync for VideoFrame {}
+
+    impl Drop for VideoDecoder {
+        fn drop(&mut self) {
+            self.decoder.close();
+        }
+    }
+
+    impl VideoDecoder {
+        pub fn new(render_context: &RenderContext, data: VideoData) -> Option<Self> {
+            let frames = Arc::new(Mutex::new(Vec::with_capacity(16)));
+
+            let decoder = init_video_decoder({
+                let frames = frames.clone();
+                move |frame: web_sys::VideoFrame| {
+                    frames.lock().push((
+                        TimeMs::new(frame.timestamp().unwrap_or(0.0)),
+                        VideoFrame(frame),
+                    ));
+                }
+            })?;
+
+            let queue = render_context.queue.clone();
+
+            // NOTE: both textures are assumed to be rgba8unorm
+            let texture = super::alloc_video_frame_texture(
+                &render_context.device,
+                &render_context.gpu_resources.textures,
+                data.config.coded_width as u32,
+                data.config.coded_height as u32,
+            );
+            let zeroed_texture_float = GpuTexture2D::new(
+                render_context
+                    .texture_manager_2d
+                    .zeroed_texture_float()
+                    .clone(),
+            )
+            .expect("expected texture to be 2D");
+
+            let mut this = Self {
+                data,
+                queue,
+                texture,
+                zeroed_texture_float,
+
+                decoder,
+
+                frames,
+                last_used_frame_timestamp: TimeMs::new(f64::MAX),
+                current_segment_idx: usize::MAX,
+                current_sample_idx: usize::MAX,
+            };
+
+            // immediately enqueue some frames, assuming playback at start
+            this.reset();
+            let _ = this.get_frame(TimeMs::new(0.0));
+
+            Some(this)
+        }
+
+        pub fn duration_ms(&self) -> f64 {
+            self.data.duration.as_f64()
+        }
+
+        pub fn width(&self) -> u32 {
+            self.data.config.coded_width as u32
+        }
+
+        pub fn height(&self) -> u32 {
+            self.data.config.coded_height as u32
+        }
+
+        pub fn get_frame(&mut self, timestamp: TimeMs) -> GpuTexture2D {
+            if timestamp.as_f64() < 0.0 {
+                // TODO(andreas): This is a hack, we should have a better way to handle this
+                return self.zeroed_texture_float.clone();
+            }
+
+            let Some(segment_idx) =
+                latest_at_idx(&self.data.segments, |segment| segment.timestamp, &timestamp)
+            else {
+                return self.texture.clone();
```

Comment:
Shouldn't this return a zeroed texture instead, since this only happens if we're before the first segment?